jumanji/environments/logic/sudoku/utils.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.logic.sudoku.constants import BOARD_WIDTH, BOX_IDX
+
+
+def apply_action(action: chex.Array, board: chex.Array) -> chex.Array:
+    return board.at[action[0], action[1]].set(action[2])
+
+
+def puzzle_completed(board: chex.Array) -> chex.Array:
+    """Checks that every row, column and 3x3 boxes includes all figures.
+
+    Args:
+        board: The sudoku board.
+
+    Returns:
+        condition: A `bool` indicator that validates if the puzzle is solved.
+    """
+
+    def _validate_row(row: chex.Array) -> chex.Array:
+        condition = jnp.sort(row) == jnp.arange(BOARD_WIDTH)
+        return condition.all()
+
+    condition_rows = jax.vmap(_validate_row)(board).all()
+    condition_columns = jax.vmap(_validate_row)(board.T).all()
+    condition_boxes = jax.vmap(_validate_row)(jnp.take(board, jnp.array(BOX_IDX))).all()
+    return condition_rows & condition_columns & condition_boxes
+
+
+def validate_board(board: chex.Array) -> chex.Array:
+    """Checks that every row, column and 3x3 boxes contain no duplicate number.
+
+    Args:
+        board: The sudoku board.
+
+    Returns:
+        condition: A `bool` indicator that validates a board or not.
+    """
+
+    def _validate_row(row: chex.Array) -> chex.Array:
+        return jax.nn.one_hot(row, BOARD_WIDTH).sum(axis=0).max() <= 1
+
+    condition_rows = jax.vmap(_validate_row)(board).all()
+    condition_columns = jax.vmap(_validate_row)(board.T).all()
+    condition_boxes = jax.vmap(_validate_row)(jnp.take(board, jnp.array(BOX_IDX))).all()
+
+    return condition_rows & condition_columns & condition_boxes
+
+
+def update_action_mask(action_mask: chex.Array, board: chex.Array) -> chex.Array:
```

Comment:
Intuitively, it seems that the action mask should only depend on the board. Is it necessary to start from the old action mask ?

Reply:
You're right I'm updating it